service/history/historyEngine.go

```diff
@@ -2557,6 +2560,526 @@ func (e *historyEngineImpl) SyncActivity(ctx context.Context, request *h.SyncAct
 	return e.replicator.SyncActivity(ctx, request)
 }
 
+// ResetWorkflowExecution only allows resetting to decisionTaskCompleted, but exclude that batch of decisionTaskCompleted.
+// It will then fail the decision with cause of "reset_workflow"
+func (e *historyEngineImpl) ResetWorkflowExecution(ctx context.Context, resetRequest *h.ResetWorkflowExecutionRequest) (response *workflow.ResetWorkflowExecutionResponse, retError error) {
+	domainEntry, retError := e.getActiveDomainEntry(resetRequest.DomainUUID)
+	if retError != nil {
+		return
+	}
+	domainID := domainEntry.GetInfo().ID
+
+	request := resetRequest.ResetRequest
+	if request == nil || request.WorkflowExecution == nil || len(request.WorkflowExecution.GetRunId()) == 0 || len(request.WorkflowExecution.GetWorkflowId()) == 0 {
+		retError = &workflow.BadRequestError{
+			Message: fmt.Sprintf("Require workflowId and runId."),
+		}
+		return
+	}
+	forkExecution := workflow.WorkflowExecution{
+		WorkflowId: request.WorkflowExecution.WorkflowId,
+		RunId:      request.WorkflowExecution.RunId,
+	}
+	newRunID := uuid.New()
+	response = &workflow.ResetWorkflowExecutionResponse{
+		RunId: common.StringPtr(newRunID),
+	}
+
+	forkContext, forkRelease, retError := e.historyCache.getOrCreateWorkflowExecutionWithTimeout(ctx, domainID, forkExecution)
+	if retError != nil {
+		return
+	}
+	defer func() { forkRelease(retError) }()
+	forkMutableState, retError := forkContext.loadWorkflowExecution()
+	if retError != nil {
+		return
+	}
+
+	// also load the current run of the workflow, it can be different from the forking runID
+	resp, retError := e.shard.GetExecutionManager().GetCurrentExecution(&persistence.GetCurrentExecutionRequest{
```

Comment:
This file is already getting large.  Can you move reset logic to its own component, so it is easy to maintain and test.

Reply:
OK